src/interchange/src/avro.rs

```diff
@@ -1988,22 +1985,39 @@ impl SchemaCache {
     /// Looks up the writer schema for ID. If the schema is literally identical
     /// to the reader schema, as determined by the reader schema fingerprint
     /// that this schema cache was initialized with, returns None.
-    async fn get(&mut self, id: i32, reader_schema: &Schema) -> anyhow::Result<Option<&Schema>> {
+    async fn get(&mut self, id: i32, reader_schema: &Schema) -> anyhow::Result<&Schema> {
         match self.cache.entry(id) {
-            Entry::Occupied(o) => Ok(o.into_mut().as_ref()),
+            Entry::Occupied(o) => o
+                .into_mut()
+                .as_ref()
+                .map_err(|e| anyhow::Error::new(e.clone())),
             Entry::Vacant(v) => {
-                // TODO(benesch): make this asynchronous, to avoid blocking the
-                // Timely thread on this network request.
-                let res = self.ccsr_client.get_schema_by_id(id).await?;
-                let schema = parse_schema(&res.raw)?;
-                if schema.fingerprint::<Sha256>().bytes == self.reader_fingerprint.bytes {
-                    Ok(v.insert(None).as_ref())
-                } else {
-                    // the writer schema differs from the reader schema,
-                    // so we need to perform schema resolution.
-                    let resolved = resolve_schemas(&schema, reader_schema)?;
-                    Ok(v.insert(Some(resolved)).as_ref())
-                }
+                // An issue with _fetching_ the schema should be returned
+                // immediately, and not cached, since it might get better on the
+                // next retry.
+                // TODO - some sort of exponential backoff or similar logic
+                let response = self.ccsr_client.get_schema_by_id(id).await?;
+                // Now, we've gotten some json back, so we want to cache it (regardless of whether it's a valid
+                // avro schema, it won't change).
+                //
+                // However, we can't just cache it directly, since resolving schemas takes significant CPU work,
+                // which  we don't want to repeat for every record. So, parse and resolve it, and cache the
+                // result (whether schema or error).
+                let rf = &self.reader_fingerprint.bytes;
+                let result = (|| {
+                    let schema = Schema::parse_str(&response.raw)?;
+                    if &schema.fingerprint::<Sha256>().bytes == rf {
+                        Ok(schema)
+                    } else {
+                        // the writer schema differs from the reader schema,
+                        // so we need to perform schema resolution.
+                        let resolved = resolve_schemas(&schema, reader_schema)?;
+                        Ok(resolved)
+                    }
+                })();
+                v.insert(result)
```

Comment:
If a user's schema registry is restarted, will all schemas be repopulated with the same ids? If not, this caching might cause an issue. Maybe the items in this cache should have some sort of expiration?

Reply:
Good question, Schema IDs are immutable though. 

In fact, Confluent Schema Registry itself stores them in Kafka, for durability, in the `_schemas` topic:

```
$ kafka-console-consumer --bootstrap-server localhost:9092 --topic _schemas --from-beginning | grep test-value 
{"subject":"test-value","version":1,"id":101,"schema":"{\"type\":\"record\",\"name\":\"BtvValue\",\"fields\":[{\"name\":\"f1\",\"type\":\"long\"},{\"name\":\"f2\",\"type\":\"string\"}]}","deleted":false}
{"subject":"test-value","version":2,"id":102,"schema":"{\"type\":\"record\",\"name\":\"BtvValue\",\"fields\":[{\"name\":\"blah\",\"type\":\"long\"},{\"name\":\"f2\",\"type\":\"string\"}]}","deleted":false}
```